### PR TITLE
vmalert: support configuration file for notifiers

### PR DESCRIPTION
**Pull request by hagen1778**
* vmalert notifiers now can be configured via file
see https://docs.victoriametrics.com/vmalert.html#notifier-configuration-file
* add support of Consul service discovery for notifiers config
see https://github.com/VictoriaMetrics/VictoriaMetrics/issues/1947
* add UI section for currently loaded/discovered notifiers
* deprecate `-rule.configCheckInterval` in favour of `-configCheckInterval`

Signed-off-by: hagen1778 <roman@victoriametrics.com>

----------------------

image

### CODE REVIEW

**Review by f41gh7 — COMMENTED** [review]
Looks good, thanks! have few minor comments

File: app/vmalert/group.go [hunk 1]
```diff
@@ -247,7 +247,7 @@ func (g *Group) start(ctx context.Context, nts []notifier.Notifier, rw *remotewr
 
 	logger.Infof("group %q started; interval=%v; concurrency=%d", g.Name, g.Interval, g.Concurrency)
 	e := &executor{rw: rw}
-	for _, nt := range nts {
+	for _, nt := range nts() {
 		ent := eNotifier{
 			Notifier:         nt,
 			alertsSent:       getOrCreateCounter(fmt.Sprintf("vmalert_alerts_sent_total{addr=%q}", nt.Addr())),
```
[the remark this answers is not in the document]
Comment (hagen1778) [reply app/vmalert/group.go]: my bad... totally forgot about this. Please see e10b1ec26524ed97c67a0a812a71589ece0fb85f

File: app/vmalert/notifier/config.go [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+	if err != nil {
+		return nil, fmt.Errorf("cannot obtain abs path for %q: %w", path, err)
+	}
+	cfg.baseDir = filepath.Dir(absPath)
+	return cfg, nil
+}
+
+func parseLabels(target string, metaLabels map[string]string, cfg *Config) (string, []prompbmarshal.Label, error) {
+	labels := mergeLabels(target, metaLabels, cfg)
+	relabelConfigs, err := promrelabel.ParseRelabelConfigs(cfg.RelabelConfigs, false)
```
[the remark this answers is not in the document]
Comment (hagen1778) [reply app/vmalert/notifier/config.go]: Good point!

File: app/vmalert/notifier/config_watcher.go [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+		target := labels["__address__"]
+		u, processedLabels, err := parseLabels(target, labels, cfg)
+		if err != nil {
+			errors = append(errors, err)
+			continue
+		}
+
+		// check for duplicates
+		if _, ok := duplicates[u]; ok {
+			errors = append(errors, fmt.Errorf("duplicated url %q detected", u))
```
[the remark this answers is not in the document]
Comment (hagen1778) [reply app/vmalert/notifier/config_watcher.go]: See 1e063fc545da370def021f5e1edfddaa24dcce16

File: app/vmalert/utils/metrics.go [hunk 4]
```diff
@@ -0,0 +1,54 @@
+package utils
+
+import "github.com/VictoriaMetrics/metrics"
+
+type namedMetric struct {
+	Name string
+}
+
+// Unregister removes the metric by name from default registry
+func (nm namedMetric) Unregister() {
```
Comment (f41gh7) [inline app/vmalert/utils/metrics.go:10]: It seems, that Unregister method can be added to the metrics package for each object, it improves user experience @valyala

**Review by f41gh7 — APPROVED** [review]
LGTM

File: app/vmalert/utils/metrics.go (hunk 4 above)
Comment (hagen1778) [reply app/vmalert/utils/metrics.go:10]: Created a FR https://github.com/VictoriaMetrics/metrics/issues/32

**Pull request merged**